### PR TITLE
Fix bugs caused by improper handling of TMP/TMPDIR environment variables

### PR DESCRIPTION
**Pull request by mebeim**
TL;DR: this PR fixes bugs in `epub2txt_do_file()` which cause stack-based buffer overflow and command injection due to improper handling of TMP/TMPDIR environment variables.

---

## Bugs

There are multiple bugs in the [function `epub2txt_do_file()` in `src/epub2txt.c`](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L393):

1. The environment variables `TMP` and `TMPDIR` read through `getenv()` are copied as is inside fixed-size buffers on the stack using `strcpy()` ([lines 409 and 411](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L409-L411)) and `sprintf()` (lines [450](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L450), [457](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L457), [464](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L464), [493](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L493), [503](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L503)), which uses the same buffers to make additional copies. This can easily cause a stack buffer overflow and crash the program in case the path stored in `TMP` or `TMPDIR` is longer than the destination buffer size.
2. The same environment variables are used to execute the `chmod` and `rm` commands through the `system()` function (line [451](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L452) and [505](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L505)), without any escaping. If the path specified in those variables contains double quote characters (`"`) this breaks the functionality of the program as the command becomes invalid, and can also result in executing unwanted commands (command injection).

The above bugs can also be considered security vulnerabilities, depending on the scenario, as they can lead to the execution of arbitrary code/commands in the context of the application. If we imagine a scenario where this tool is used to convert EPUBs through a cgi-bin web server where users might have partial control over environment variables for temporary directories, this could also result in remote code execution.

Example 1:

```none
$ export TMP=$(printf '%0800d' 0)
$ epub2txt blank.epub
checkdir:  cannot create extraction directory: 000000000000000000000[...]/epub2txt16058
           File name too long
sh: 1: Syntax error: Unterminated quoted string
rm: cannot remove '000000000000000000000[...]rm -rf ': File name too long
rm: cannot remove '000000000000000000000[...]chmod -R 744 000000000000000000000[...]/epub2txt16058': File name too long
Segmentation fault
```

Example 2:

```none
$ export TMP='";echo THIS IS A COMMAND INJECTION;exit 0;"'
$ epub2txt blank.epub
checkdir:  cannot create extraction directory: ";echo THIS IS A COMMAND INJECTION;exit 0;"/epub2txt15941
           No such file or directory
chmod: cannot access '': No such file or directory
THIS IS A COMMAND INJECTION
THIS IS A COMMAND INJECTION
epub2txt: Can't open file '";echo THIS IS A COMMAND INJECTION;exit 0;"/epub2txt15941/META-INF/container.xml' for reading: No such file or directory
```

Note: as these issues are independent of the EPUB being parsed, the `blank.epub` file used for testing above just contains one blank page.

## Fixes

The buffer overflow was solved by simply dynamycally allocating strings with `asprintf()` instead of using fixed-size stack buffers. The second issue of passing unescaped paths to `system()` was solved by adding an helper function (file `src/util.c`) which uses `fork()` + `execvp()` to run helper commands. See also the commit messages for more info.

Additionally, I also added a simple check on the [the `mkdir()` call](https://github.com/kevinboone/epub2txt2/blob/973ffba2f4d6adbc8d1d740ccc878ff14170b1db/src/epub2txt.c#L424) to return early in case of failure, since there isn't much else to do except bail out. Continuing to run code that tries to operate on a non-existent directory isn't going to work anyway.

### CODE REVIEW

**Comment by kevinboone** [comment]
I've agonized over this for a long time. You say:

"If we imagine a scenario where this tool is used to convert EPUBs through a cgi-bin web server..."

My simple solution is: don't do this :) I wrote epub2txt primarily to read EPUBs on minimal/embedded Linux systems (including Android). This is clear in the documentation. I never intended it to be used in a hostile, Internet-facing environment, and it really isn't suitable for that application.

I have fixed a number of problems related to bad input (defective EPUBs, etc); but each fix adds a bit of weight to the program. I imagine that there are many other ways in which epub2txt could be subverted, given the right input. To be honest, I'm not very keen on continuing to pursue this kind of thing, because the program is getting larger and larger.

I think it's perfectly reasonable that somebody could want an EPUB-to-text converter that _is_ suitable for use in hostile environments, but I don't have the resources to maintain one, alongside this one. I'm very happy to collaborate with somebody who does want to do this.

I will be adding a section to the documentation, explaining that epub2txt does not have security as a top priority, and that it isn't designed to handle hostile input.

Comments welcome.

**Comment by mebeim** [comment]
@kevinboone I see that, however that part of my comment was merely a "note". I am not saying that this needs to be fixed because it is necessarily a security vulnerability, just noting that it could be considered one. 

My patch here is just meant to fix existing bugs that cause the program to crash in legitimate usage case scenarios. It'd be perfectly valid for somebody to use the tool with a long temporary directory path, and I don't think that should break it to be honest. Same reasoning goes for the command injection. Paths can contain `"` characters, and that should not break the program. At the end of the day, that's pretty much all there's to it.

**Comment by kevinboone** [comment]
You're quite right. Security aside, it's legitimate for paths to contain quotes and, frankly, my use of system() to launch platform commands is just lazy. Thank you for the contribution.
